optimize-nextflow.py

```diff
@@ -0,0 +1,177 @@
+#!/usr/bin/env python3
+
+import json
+import shutil
+from collections import namedtuple
+from dataclasses import astuple, dataclass
+from functools import cached_property
+from math import ceil
+from pathlib import Path
+from textwrap import dedent, indent
+from typing import Generic, TypeVar
+
+import typer
+
+Number = TypeVar("Number", int, float)
+
+app = typer.Typer(
+    context_settings={"max_content_width": shutil.get_terminal_size().columns}
+)
+
+
+@app.command()
+def from_json(metrics_path: Path):
+    """Process the metrics JSON output from the Tower CLI.
+
+    Here's an example Tower CLI command:
+
+        tw --output "json" runs view -w "<org>/<workspace>" -i "<workflow-id>" metrics
+    """
+    json_parser = MetricsJsonParser(metrics_path)
+    task_metrics = json_parser.parse()
+    config_generator = NextflowConfigGenerator(task_metrics)
+    config = config_generator.generate()
+    print(config)
+
+
+@app.command()
+def from_api():
+    """Process the metrics JSON output from the Tower API.
+
+    Not implemented yet.
+    """
+    ...
+
+
+@dataclass
+class Stats(Generic[Number]):
+    q1: Number
+    q2: Number
+    q3: Number
+    min: Number
+    max: Number
+    mean: Number
+
+    @cached_property
+    def max_q3_ratio(self):
+        return round(self.max / self.q3, 2)
+
+
+@dataclass
+class TaskMetrics:
+    name: str
+    mem: Stats
+    mem_pct: Stats
+    cpu: Stats
+    cpu_pct: Stats
+
+    @staticmethod
+    def bytes_to_gb(memory: int) -> int:
+        memory_gb = memory / 1024**3
+        return ceil(memory_gb + 0.5)
+
+    @staticmethod
+    def load_to_cores(load: float) -> int:
+        cores = load / 100
+        return ceil(cores + 0.5)
+
+    @cached_property
+    def mem_gb(self) -> Stats:
+        values = astuple(self.mem)
+        values_adj = [self.bytes_to_gb(x) for x in values]
+        return Stats(*values_adj)
+
+    @cached_property
+    def cpu_cores(self) -> Stats:
+        values = astuple(self.cpu)
+        values_adj = [self.load_to_cores(x) for x in values]
+        return Stats(*values_adj)
+
+    def optimize_mem(self) -> int:
+        if self.mem.max == 1 and self.mem_pct.max <= 80:
+            return self.mem_gb.max
+        elif self.mem_gb.max <= 8:
+            return self.mem_gb.max
+        elif self.mem_gb.max_q3_ratio <= 1.5 and self.mem_gb.max <= 32:
+            return self.mem_gb.max
+        elif self.mem_gb.max_q3_ratio <= 1.2 and self.mem_gb.max <= 64:
+            return self.mem_gb.max
```

Comment:
Nit: For readability, I like what you've done here, but this does seem like 4 conditions that can be done with an `or`.  

Can you speak to how you chose each of these conditions?

Reply:
I intentionally kept the conditions separate to improve readability.  As far as I can tell, combining them using `or` doesn't provide a benefit that outweighs the loss in readability. Am I missing an advantage of combining the conditions? 

As for the specific conditions, they're rough heuristics. There's nothing scientific about them. In general, if the memory usage is low, I'm more willing to adjust the memory allocation to the max memory usage. If there's a big gap between Q3 and max, then I use a combination of the ratio and the max memory usage to decide whether to set the memory allocation to the max or Q3. I originally had comments describing the conditions using English, but I think the Python syntax is more succinct.

Reply:
Agreed. Can you add a comment about how these are rough heuristics?

Reply:
Yup, I mentioned that these are heuristics in the docstrings.